Error handling

IharBury opened the pull request:
Projectors include information about the currently projected events and transactions in the exception.
Dispatcher handles exceptions by logging them and cancelling the event subscription.

IharBury commented:
Related to #35

dennisdoomen commented:
Thought about this a bit more. I think this should work like this:

* If a projector or child-projector detects an exception while processing an event, it will wrap the caught exception in a `ProjectionException` which takes a message, the caught exception, a reference to the current exception and a reference to the current transaction.
* Any parent projector will just pass this exception to its parent projector (formerly known as *bootstrapper*)
* The `ProjectionException` is not tight to the specific projector but is defined by the main project. This also allows upcoming features like the `StatisticsService` to collect them for diagnostic purposes
* If the `Dispatcher` detects such an exception, it will log

IharBury commented:
@dennisdoomen, I've reimplemented the PR according to the proposal

dennisdoomen commented on Src/LiquidProjections.NHibernate/NHibernateChildProjector.cs:
```diff
@@ -34,7 +35,29 @@ public sealed class NHibernateChildProjector<TProjection, TKey> : INHibernateChi
+            }
+            catch (Exception exception)
+            {
+                ProjectionException.ThrowNewOrEnrichExisting(exception, projectionException =>
```
Why so complicated and just throw the exception and passing in `exception` and the current `event`?

dennisdoomen commented on Src/LiquidProjections.NHibernate/NHibernateChildProjector.cs:
```diff
@@ -34,7 +35,29 @@ public sealed class NHibernateChildProjector<TProjection, TKey> : INHibernateChi
+            {
+                ProjectionException.ThrowNewOrEnrichExisting(exception, projectionException =>
+                {
+                    projectionException.ChildProjector = typeof(TProjection).ToString();
```
Isn't this obvious from the stack trace?

IharBury replied:
Stack trace does not contain generic parameters

dennisdoomen commented on Src/LiquidProjections/Dispatcher.cs:
```diff
@@ -18,27 +20,37 @@ public Dispatcher(IEventStore eventStore)
+                    }
+                    catch (Exception exception)
+                    {
+                        ProjectionException.LogException(exception, LogLevel.Fatal);
```
I don't think the exception itself should log directly. That's mixing responsibilities

dennisdoomen commented on Src/LiquidProjections/ProjectionException.cs:
```diff
@@ -0,0 +1,375 @@
+        /// <param name="setProperties">
+        /// The delegate that enriches the provided <see cref="ProjectionException"/> with additional data.
+        /// </param>
+        public static void ThrowNewOrEnrichExisting(Exception innerException, Action<ProjectionException> setProperties)
```
Why do you need this factory here instead of having the throwing party create an instance and throw it explicitly?

dennisdoomen commented on Src/LiquidProjections/ProjectionException.cs:
```diff
@@ -0,0 +1,375 @@
+
+            Exception currentException = exception;
+
+            while (currentException != null)
```
I think this functionally belongs in the hosting code (e.g. VisionSuite). The exception should only expose a method of some sort that will give you a nice JSON representation of the transaction batch, its events and the specific event that triggered it.

IharBury replied:
I don't think we need JSON conversion in LiquidProjections at all then. Exposing the objects themselves from the exception should be enough.

dennisdoomen commented on Tests/LiquidProjections.NHibernate.Specs/NHibernateProjectorSpecs.cs:
```diff
@@ -1197,6 +1195,43 @@ private class ChildProjectionState
+            [Fact]
+            public async Task Then_it_should_throw_projection_exception()
+            {
+                ProjectionException exception = await Assert.ThrowsAsync<ProjectionException>(
```
It's weird not to have a `When`. You can employ Chill's [result-type mechanism](https://github.com/dennisdoomen/VisionSuite/commit/2facc230c0569848e0d7dac969c89ebc2d7e2d01#diff-6541c4c2ab18df4686ce3262c414979cR216) to support this. Same for the RavenDB specs.

dennisdoomen commented on Tests/LiquidProjections.NHibernate.Specs/NHibernateProjectorSpecs.cs:
```diff
@@ -1197,6 +1195,43 @@ private class ChildProjectionState
+                ProjectionException exception = await Assert.ThrowsAsync<ProjectionException>(
+                    () => The<MemoryEventSource>().Write(UseThe(new CategoryDiscontinuedEvent())));
+
+                exception.Projector.Should().Be(typeof(ProductCatalogEntry).ToString());
```
I would probably split the current event, the transaction batch and the innerexception into three *facts*. You can also make the tests more readable by comparing the properties with `ShouldBeEquivalentTo`. Same for the RavenDB specs.

IharBury commented:
I've addressed the review comments. I've also found 3 bugs in Chill so waiting for those PRs to be merged and the new version to be released to avoid creating work-arounds.

IharBury commented:
And it seems that RavenDB specs tend to hang when run together. I don't know the exact reason yet. Maybe it's related to Chill doing `Wait` for async specs... It may deadlock with RavenDB asyncs somehow...

IharBury commented:
I had to make the tests run synchronously to avoid the RavenDB specs hanging
